### ISSUE TITLE
Developing Custom Commands - any additional documentation?

### ISSUE THREAD

**Issue by edwardmsmith** [tiff/wysihtml5#43]
Is there any more documentation or resources on writing custom commands beyond the source of the existing commands (and commands.js - which is pretty nicely documented, by the way!)

I need to write a sort of oEmbed command, and want to use the built-in dialog handling (round-trip - using the dialog to edit existing elements, as well as new elements)

The comments in the code are pretty good, but wanted to check, before I got going, if there was any additional information on best ways to do this sort of thing.

Thanks!

**Comment by tiff** [comment]
Hi,

so far there's no documentation on how to write custom commands but it should be pretty straight forward:

Example:

``` javascript
wysihtml5.commands.insertFoo = {
  // exec usually behaves like a toggle
  // if the format is applied then undo it (and vica versa)
  exec: function(composer, command, param) {
    if (this.state(composer, command, param) {
      // undo code here
    } else {
      // apply code here (eg. insert <foo>)
    }
  },

  // usually returns a truthy value when the command is applied to the current selection
  // a falsy when the current selection isn't formatted with <foo>
  state: function(composer, command) {
     return true;
  },

  // ignore this for now (it's currently not used)
  value: function() {
  }
};
```

Also you might wanna check out the [documentation and source of all currently available commands](https://github.com/xing/wysihtml5/wiki/Supported-Commands).
Consider starting with the [insertImage](https://github.com/xing/wysihtml5/blob/master/src/commands/insertImage.js) and [createLink](https://github.com/xing/wysihtml5/blob/master/src/commands/createLink.js) commands as they cover a lot of code that you could reuse and I think the code isn't that hard to read.
If you just need to wrap an element then the already existing formatInline/formatBlock commands are very helpful.

**Comment by tiff** [comment]
Can this be closed, @edwardmsmith?

**Comment by edwardmsmith** [comment]
Yessir, thanks for the info.

**Comment by ra00f** [comment]
Hello,

Sorry open it again, I don't understand the logic to make additional buttons. 

I need to make buttons that would insert some uncode symbols to the typing area, when clicked. 

I've created the front end part, but no idea how to make the backend.

I've created a new ticket at https://github.com/xing/wysihtml5/issues/176. I also have a link to my demo page at http://itsnot.freehostia.com/wysihtml5/wysihtml5-new.htm

How do I edit the script to make the buttons to insert fractions and the other symbols work?

Thanks

**Comment by tmaiaroto** [comment]
I'm also stuck. "commands" is apparently undefined. I see the constructor and in the console I'm drilling down and seeing it, but I can't set the property. It just says can not set property of undefined.

I went to the page @ra00f linked and was able to insert things via his function, though only after having something typed in the editor. Maybe I'll start there to see if I can make that a little smoother.

...But it would be really nice if we could extend this really awesome editor. I really want to just have modals that allow me to do some jazz and then insert some end result string into the editor.

**Comment by tmaiaroto** [comment]
If you put the code in for a dialog with insertHTML (much like you would for createLink) you can get a dialog with input field... The only problem is it sends [object Object] to the editor. If this were to change (or there was a new command) that would look at the input field and instead return the val() then I think you'd be golden.

...And with that, you could essentially have whatever logic need be completely separate from the editor and people could allow image uploads or whatever other stuff they need in that modal and then set the final value to the input field (which perhaps may be hidden) or just leave it as an open input field (much like createLink) or even a textarea. 

Then I think you can essentially have all the flexibility one could want... I mean a more formal plugin system would be nice, but if you can just pass your now very custom commands to the "toolbar" property...I think that would be enough.
